README.md
# ![TypeScript for Unreal (Examples)][bnr]

A couple of examples showcasing [TypeScript for Unreal][tsu].

Need help with something? Join us on [Discord][dsc].

This project is based on the Blueprint Input level in the [Content Examples][cex] project and is meant to show the difference between making something in Blueprint versus making something in TypeScript for Unreal.

## Downloads

Please note that **currently**...

- TSU is in **alpha**
- TSU has **only** been tested with **Unreal Engine 4.21**
- You can **not** use TSU outside of Win64 **editor** builds
- You can **not** use TSU in packaged/cooked builds
- There **will** be breaking changes in upcoming versions

Hopefully all of these issues will be resolved soon, but please be aware of them before downloading.

If you still want to try it, head over to [Releases][rls].

## License

This project is licensed under the 3-clause BSD license. See the [LICENSE][lic] file for details.

[bnr]: https://user-images.githubusercontent.com/4884246/54883366-87e36180-4e65-11e9-8bc9-5fdb6b5cd462.png
[cex]: https://docs.unrealengine.com/en-US/Resources/ContentExamples
[dsc]: https://discord.gg/QPrNpAQ
[lic]: LICENSE.md
[rls]: https://github.com/mihe/tsu-examples/releases
[tsu]: https://github.com/mihe/tsu


CHANGELOG.md
# Changelog

## [0.2.0](https://github.com/mihe/tsu-examples/compare/v0.1.0...v0.2.0) (2019-03-26)

- Updated TSU to 0.2.0
- Fixed issue with not being able to generate project files

## 0.1.0 (2019-03-25)

- Initial release


Content/Scripts/Source/Saucer/TBP_Saucer.ts
import { Actor } from 'UE/Actor';
import { BP_Saucer } from 'UE/BP_Saucer';
import { EAttachLocation } from 'UE/EAttachLocation';
import { GameplayStatics } from 'UE/GameplayStatics';
import { HitResult } from 'UE/HitResult';
import { KismetMathLibrary as KMath } from 'UE/KismetMathLibrary';
import { ParticleSystem } from 'UE/ParticleSystem';
import { Rotator } from 'UE/Rotator';
import { SceneComponent } from 'UE/SceneComponent';
import { StaticMeshActor } from 'UE/StaticMeshActor';
import { Vector } from 'UE/Vector';

const beamOpacityParam = 'Opacity';
const socketEvacZone = 'Evac zone';

export function construct(target: BP_Saucer) {
	target.beamMaterial = target.beam.createDynamicMaterialInstance(0);
}

export function enableBeam(target: BP_Saucer) {
	target.beamOn = true;
	target.beamLight.setVisibility(true);
	target.beamMaterial.setScalarParameterValue(beamOpacityParam, 0.4);
}

export function disableBeam(target: BP_Saucer) {
	target.beamOn = false;
	target.beamLight.setVisibility(false);
	target.beamMaterial.setScalarParameterValue(beamOpacityParam, 0.01);
}

export function tick(target: BP_Saucer, deltaTime: number) {
	spinTop(target, deltaTime);
	updateVelocity(target, deltaTime);
	updateSpringArm(target, deltaTime);
	updateCameraAndBody(target, deltaTime);
	updateFieldOfView(target, deltaTime);
	tryAbduct(target, deltaTime);
}

function spinTop(target: BP_Saucer, deltaTime: number) {
	target.top.addRelativeRotation(new Rotator(0, 0, 180 * deltaTime));
}

function updateVelocity(target: BP_Saucer, deltaTime: number) {
	const inputs = new Vector(
		target.axisMoveForward,
		target.axisMoveRight,
		target.axisMoveUp
	);

	const cameraYaw = target.camera.getComponentRotation().yaw;
	const newVelocity = new Rotator(0, 0, cameraYaw)
		.rotateVector(inputs)
		.multiplyFloat(target.movementSpeed * deltaTime)
		.add(target.getVelocity());

	target.body.setPhysicsLinearVelocity(newVelocity);
}

function updateSpringArm(target: BP_Saucer, deltaTime: number) {
	const {
		springArm,
		cameraSpeed,
		axisLookUp,
		axisTurn
	} = target;

	const currentRotation = springArm.getComponentToWorld().rotation;
	const currentYaw = currentRotation.yaw;
	const currentPitch = currentRotation.pitch;

	const cameraSpeedInv = cameraSpeed * -1;

	const targetYaw = currentYaw + (cameraSpeed * axisTurn);
	let targetPitch = currentPitch + (cameraSpeedInv * axisLookUp);
	targetPitch = KMath.clampAngle(targetPitch, -45, 85);

	const targetRotation = new Rotator(0, targetPitch, targetYaw);

	const newRotation = KMath.rinterpTo(
		currentRotation,
		targetRotation,
		deltaTime,
		8
	);

	springArm.setWorldRotation(newRotation);
}

function updateCameraAndBody(target: BP_Saucer, deltaTime: number) {
	const { camera } = target;
	const currentCameraTransform = camera.getComponentToWorld();
	const currentCameraLocation = currentCameraTransform.location;
	const currentCameraRotation = currentCameraTransform.rotation;

	const currentBodyTransform = target.body.getComponentToWorld();
	const currentBodyLocation = currentBodyTransform.location;
	const currentBodyRotation = currentBodyTransform.rotation;

	const targetCameraRotation = KMath.findLookAtRotation(
		currentCameraLocation,
		currentBodyLocation
	);

	const newCameraRotation = KMath.rinterpTo(
		currentCameraRotation,
		targetCameraRotation,
		deltaTime,
		2
	);

	camera.setWorldRotation(newCameraRotation);

	const targetBodyRoll = target.axisMoveRight * 20;
	const targetBodyPitch = target.axisMoveForward * -7;
	const targetBodyRotation = new Rotator(
		targetBodyRoll,
		targetBodyPitch,
		targetCameraRotation.yaw
	);

	const newBodyRotation = KMath.rinterpTo(
		currentBodyRotation,
		targetBodyRotation,
		deltaTime,
		4
	);

	target.setActorRotation(newBodyRotation);
}

function updateFieldOfView(target: BP_Saucer, deltaTime: number) {
	const speed = target.getVelocity().size();
	const targetFov = KMath.mapRangeClamped(speed,
		300,
		1000,
		75,
		110
	);

	const { camera } = target;
	camera.fieldOfView = KMath.finterpTo(
		camera.fieldOfView,
		targetFov,
		deltaTime,
		8
	);
}

function tryAbduct(target: BP_Saucer, deltaTime: number) {
	if (!target.beamOn) { return; }

	const { overlappingActors } = target.beamTrigger.getOverlappingActors();
	if (overlappingActors.length === 0) { return; }

	for (const actor of overlappingActors) {
		if (!(actor instanceof StaticMeshActor)) { continue; }

		const targetPos = target.getActorLocation();
		const actorPos = actor.getActorLocation();

		const currentVelocity = actor.getVelocity();
		const targetVelocity = targetPos
			.subtract(actorPos)
			.getSafeNormal()
			.multiply(new Vector(1, 1, 0.3))
			.multiplyFloat(target.beamStrength);

		const newVelocity = KMath.vinterpTo(
			currentVelocity,
			targetVelocity,
			deltaTime,
			5
		);

		actor.staticMeshComponent.setPhysicsLinearVelocity(newVelocity);
	}
}

export function tryBoost(target: BP_Saucer) {
	if (!target.boostReady) { return; }

	const { body, movementSpeed } = target;
	const { rotation } = body.getComponentToWorld();
	const forward = rotation.getForwardVector();
	const newVelocity = forward.withZ(0)
		.multiplyFloat(movementSpeed * 2);

	body.setPhysicsLinearVelocity(newVelocity);

	target.boostReady = false;

	spawnEmitterAttached(
		target.boostEffect,
		target.engine1,
		EAttachLocation.SnapToTarget
	);

	spawnEmitterAttached(
		target.boostEffect,
		target.engine2,
		EAttachLocation.SnapToTarget
	);

	setTimeout(() => {
		target.boostReady = true;
	}, 800);
}

export function onAbducted(target: BP_Saucer, other: Actor) {
	if (!target.beamOn) { return; }
	if (!(other instanceof StaticMeshActor)) { return; }

	other.staticMeshComponent.setSimulatePhysics(false);
	other.staticMeshComponent.setVisibility(false);
	other.setActorEnableCollision(false);
	target.abducted.push(other);

	spawnEmitterAttached(
		target.abductEffect,
		target.body,
		EAttachLocation.KeepRelativeOffset,
		new Vector(0, 0, -10)
	);
}

export function tryEject(target: BP_Saucer) {
	const { abducted } = target;

	const subject = abducted.pop();
	if (!subject) { return; }

	interruptBeam(target);

	const { body } = target;
	const evacLocation = body.getSocketLocation(socketEvacZone);

	subject.staticMeshComponent.setSimulatePhysics(true);
	subject.setActorLocation(evacLocation);
	subject.setActorEnableCollision(true);
	subject.staticMeshComponent.setVisibility(true);

	const subjectVelocity = target
		.getActorRotation()
		.rotateVector(new Vector(0, 0, -1500));

	subject.staticMeshComponent.setPhysicsLinearVelocity(subjectVelocity);
	body.setPhysicsLinearVelocity(subjectVelocity.multiplyFloat(-0.1));

	spawnEmitterAttached(
		target.ejectEffect,
		body,
		EAttachLocation.KeepRelativeOffset,
		new Vector(0, 0, -10)
	);
}

function interruptBeam(target: BP_Saucer) {
	disableBeam(target);
	setTimeout(() => {
		if (target.wantsBeam) {
			enableBeam(target);
		}
	}, 500);
}

export function onImpact(target: BP_Saucer, hit: HitResult) {
	if (target.getVelocity().size() >= 300) {
		spawnEmitterAtLocation(target.impactEffect, hit.impactPoint);
	}
}

function spawnEmitterAttached(
	template: ParticleSystem,
	attachTo: SceneComponent,
	locationType = EAttachLocation.KeepRelativeOffset,
	location = new Vector(0, 0, 0)
) {
	return GameplayStatics.spawnEmitterAttached(
		template,
		attachTo,
		'',
		location,
		new Rotator(0, 0, 0),
		new Vector(1, 1, 1),
		locationType,
		true
	);
}

function spawnEmitterAtLocation(template: ParticleSystem, location: Vector) {
	return GameplayStatics.spawnEmitterAtLocation(
		template,
		location,
		new Rotator(0, 0, 0),
		new Vector(1, 1, 1),
		true
	);
}


Content/Scripts/Source/Shooter/TBP_PixelGame.ts
import { Actor } from 'UE/Actor';
import { BP_PixelGame } from 'UE/BP_PixelGame';
import { BP_PixelShipEnemy } from 'UE/BP_PixelShipEnemy';
import { EViewTargetBlendFunction } from 'UE/EViewTargetBlendFunction';
import { GameplayStatics } from 'UE/GameplayStatics';
import { KismetMathLibrary as KMath } from 'UE/KismetMathLibrary';
import { KismetSystemLibrary as KSystem } from 'UE/KismetSystemLibrary';
import { TimelineComponent } from 'UE/TimelineComponent';
import { Transform } from 'UE/Transform';
import { Vector } from 'UE/Vector';

const cameraShakeTrack = 'Offset';

export function construct(target: BP_PixelGame) {
	target.playerSpawnPoint.setRelativeLocation(target.playerSpawnLocation);
	target.enemySpawnPoint.setRelativeLocation(target.enemySpawnLocation);
	target.trigger.setRelativeLocation(target.triggerLocation);
}

export function onBeginPlay(target: BP_PixelGame) {
	const cameraShake = new TimelineComponent(target);
	cameraShake.registerComponent();
	cameraShake.setTimelineLength(0.4);
	cameraShake.addVectorCurve(cameraShakeTrack, target.cameraShakeCurve);
	target.addOwnedComponent(cameraShake);
	target.cameraShake = cameraShake;
}

export function startGame(target: BP_PixelGame) {
	target.spawnInterval = 4;
	target.spawningEnemies = true;
	target.nextSpawnTime = KSystem.getGameTimeInSeconds() + target.spawnInterval;

	const controller = GameplayStatics.getPlayerController(0);
	const spawnTransform = target.playerSpawnPoint.getComponentToWorld();
	const playerShip = target.spawnPlayer(spawnTransform);

	playerShip.onDestroyed.add(() => onPlayerDestroyed(target));

	controller.possess(playerShip);

	controller.setViewTargetWithBlend(
		target,
		1,
		EViewTargetBlendFunction.VTBlend_EaseInOut,
		3,
		true);

	target.playerShip = playerShip;
	target.canReset = true;
}

export function quitGame(target: BP_PixelGame) {
	target.canReset = false;

	const { playerShip } = target;
	if (playerShip && playerShip.isValid()) {
		playerShip.destroyActor();
	}
}

function resetGame(target: BP_PixelGame) {
	for (const enemy of target.spawnedEnemies) {
		if (enemy && enemy.isValid()) {
			enemy.destroyActor();
		}
	}

	if (target.canReset) {
		startGame(target);
	}
}

function playCameraShake(target: BP_PixelGame, intensity: number) {
	const { cameraShake, camera } = target;

	cameraShake.setOnVectorUpdate(cameraShakeTrack, value => {
		camera.setRelativeLocation(value.multiplyFloat(intensity));
	});

	cameraShake.playFromStart();
}

export function tick(target: BP_PixelGame) {
	trySpawnEnemy(target);
}

function trySpawnEnemy(target: BP_PixelGame) {
	if (!target.spawningEnemies) { return; }

	const currentTime = KSystem.getGameTimeInSeconds();
	if (currentTime < target.nextSpawnTime) { return; }

	const spawnTransform = target.enemySpawnPoint.getComponentToWorld();

	const spawnOffsetY = KMath.randomFloatInRange(
		target.maxHorizontalSpawnOffset,
		target.maxHorizontalSpawnOffset * -1
	);

	const spawnOffset = target.getActorRotation().rotateVector(
		new Vector(0, spawnOffsetY, 0)
	);

	const enemyTransform = new Transform(
		spawnTransform.location.add(spawnOffset),
		spawnTransform.rotation,
		new Vector(1, 1, 1)
	);

	const enemy = target.spawnEnemy(enemyTransform);

	enemy.onDestroyed.add(() => {
		playCameraShake(target, 3);
	});

	target.spawnedEnemies.push(enemy);

	const spawnDelay = KMath.randomFloatInRange(
		target.spawnInterval,
		target.spawnInterval + 2
	);

	target.nextSpawnTime = currentTime + spawnDelay;
	target.spawnInterval = KMath.fmax(target.spawnInterval - 0.3, 0.7);
}

function onPlayerDestroyed(target: BP_PixelGame) {
	target.spawningEnemies = false;

	const controller = GameplayStatics.getPlayerController(0);
	controller.setViewTargetWithBlend(target);

	resetGame(target);
}

export function onBottomOverlap(other: Actor) {
	if (other instanceof BP_PixelShipEnemy) {
		other.destroyActor();
	}
}


Content/Scripts/Source/Platformer/TBP_PixelDude.ts
import { BP_PixelDude } from 'UE/BP_PixelDude';
import { Vector } from 'UE/Vector';

export function construct(target: BP_PixelDude) {
	target.initialCharacterTransform = target.getTransform();
}

export function onInput(target: BP_PixelDude, right: number, up: number) {
	if (target.deactivate) { return; }

	if (right !== 0) {
		target.addMovementInput(new Vector(1, 0, 0), right);
	}

	if (up > 0) {
		target.jump();
		target.characterBody.setStaticMesh(target.meshJumping);
	}
}

export function onLanded(target: BP_PixelDude) {
	target.characterBody.setStaticMesh(target.meshDefault);
}

export function onQuitGame(target: BP_PixelDude) {
	target.deactivate = true;
	target.characterBody.setStaticMesh(target.meshDefault);
	target.setActorTransform(target.initialCharacterTransform);
}


Content/Scripts/Source/Shooter/TBP_PixelProjectile.ts
import { Actor } from 'UE/Actor';
import { BP_PixelProjectile } from 'UE/BP_PixelProjectile';
import { GameplayStatics as KGameplay } from 'UE/GameplayStatics';
import { Rotator } from 'UE/Rotator';
import { Vector } from 'UE/Vector';

export function construct(target: BP_PixelProjectile) {
	target.projectileMaterial = (
		target.projectile.createDynamicMaterialInstance(0)
	);
}

export function onBeginPlay(target: BP_PixelProjectile) {
	const { projectileMaterial, projectileColor, light } = target;
	projectileMaterial.setVectorParameterValue('Color2', projectileColor);
	light.setLightColor(projectileColor);
}

export function move(target: BP_PixelProjectile, deltaSeconds: number) {
	target.addActorLocalOffset(
		target.speed.multiplyFloat(deltaSeconds)
	);
}

export function onImpact(target: BP_PixelProjectile, other: Actor) {
	const { projectileOwner } = target;
	if (projectileOwner && projectileOwner.equals(other)) { return; }

	other.applyDamage(20, null, null, null);

	KGameplay.spawnEmitterAtLocation(
		target.impactEffect,
		target.getActorLocation(),
		new Rotator(0, 0, 0),
		new Vector(1, 1, 1)
	);

	target.destroyActor();
}


Content/Scripts/Source/Shooter/TBP_PixelShipPlayer.ts
import { BP_PixelShipPlayer } from 'UE/BP_PixelShipPlayer';
import { GameplayStatics } from 'UE/GameplayStatics';
import { KismetMathLibrary as KMath } from 'UE/KismetMathLibrary';
import { Rotator } from 'UE/Rotator';
import { TimelineComponent } from 'UE/TimelineComponent';
import { Vector } from 'UE/Vector';

const trackName = 'Movement';

export function onBeginPlay(target: BP_PixelShipPlayer) {
	target.gameCabinetYaw = target.getInstigator().getActorRotation().yaw;
	target.spawnLocation = target.getActorLocation();

	const spawnAnimation = new TimelineComponent(target);
	spawnAnimation.registerComponent();
	spawnAnimation.setTimelineLength(1);
	spawnAnimation.addFloatCurve(trackName, target.spawnAnimationCurve);

	spawnAnimation.setOnFloatUpdate(trackName, value => {
		onAnimate(target, value);
	});

	spawnAnimation.setOnFinished(() => {
		target.movementEnabled = true;
	});

	spawnAnimation.playFromStart();

	target.addOwnedComponent(spawnAnimation);
}

function onAnimate(target: BP_PixelShipPlayer, alpha: number) {
	const { spawnLocation, gameCabinetYaw } = target;

	const sourceLocation = spawnLocation;
	const targetLocation = sourceLocation.add(new Vector(0, 0, 60));
	const location = KMath.vlerp(sourceLocation, targetLocation, alpha);
	target.setActorLocation(location);

	const sourceRotation = new Rotator(0, 0, gameCabinetYaw + 720);
	const targetRotation = new Rotator(0, 0, gameCabinetYaw);
	const rotation = KMath.rlerp(sourceRotation, targetRotation, alpha, false);
	target.setActorRotation(rotation);
}

export function tick(target: BP_PixelShipPlayer) {
	if (!target.movementEnabled) { return; }

	const deltaSeconds = GameplayStatics.getWorldDeltaSeconds();

	const step = target.axisMoveRight * target.movementSpeed;
	const desiredSpeed = step * deltaSeconds;
	target.speed = KMath.finterpTo(
		target.speed,
		desiredSpeed,
		deltaSeconds,
		4
	);

	const velocity = new Vector(target.speed, 0, 0);
	const cabinetRotation = new Rotator(0, 0, target.gameCabinetYaw);
	const offset = cabinetRotation.rotateVector(velocity);

	target.addActorWorldOffset(offset, true);

	const yaw = target.gameCabinetYaw + (target.speed * -5);
	const targetRotation = new Rotator(0, 0, yaw);
	const currentRotation = target.getActorRotation();
	const newRotation = KMath.rinterpTo(
		currentRotation,
		targetRotation,
		deltaSeconds,
		20
	);

	target.setActorRotation(newRotation);
}

export function startFiring(target: BP_PixelShipPlayer) {
	if (!target.movementEnabled) { return; }

	fireProjectile(target);

	target.fireTimerHandle = setInterval(() => {
		fireProjectile(target);
	}, 200);
}

function fireProjectile(target: BP_PixelShipPlayer) {
	target.spawnProjectile(
		target.getTransform(),
		new Vector(0, 0, 700),
		target,
		KMath.makeColor(0.2, 0.5, 1, 1)
	);
}

export function stopFiring(target: BP_PixelShipPlayer) {
	const { fireTimerHandle } = target;
	if (fireTimerHandle.isValid()) {
		clearInterval(fireTimerHandle);
	}
}

export function onDamage(target: BP_PixelShipPlayer, damage: number) {
	target.speed *= -0.5;
	target.health -= damage;

	if (target.health <= 0) {
		GameplayStatics.spawnEmitterAtLocation(
			target.destroyedEffect,
			target.getActorLocation(),
			new Rotator(0, 0, 0),
			new Vector(1, 1, 1)
		);

		target.destroyActor();
	}
}


Content/Scripts/Source/Shooter/TBP_PixelShipEnemy.ts
import { Actor } from 'UE/Actor';
import { BP_PixelShipEnemy } from 'UE/BP_PixelShipEnemy';
import { GameplayStatics } from 'UE/GameplayStatics';
import { KismetMathLibrary as KMath } from 'UE/KismetMathLibrary';
import { KismetSystemLibrary as KSystem } from 'UE/KismetSystemLibrary';
import { Pawn } from 'UE/Pawn';
import { Rotator } from 'UE/Rotator';
import { TimelineComponent } from 'UE/TimelineComponent';
import { Vector } from 'UE/Vector';

const trackName = 'Movement';

export function onBeginPlay(target: BP_PixelShipEnemy) {
	target.gameCabinetYaw = target.getInstigator().getActorRotation().yaw;
	target.movementDirection = KMath.randomBool();
	target.movementSpeed = KMath.randomFloatInRange(25, 200);
	target.spawnLocation = target.getActorLocation();

	const projectileDelay = KMath.randomFloatInRange(0.4, 0.8);
	target.nextProjectileTime = KSystem.getGameTimeInSeconds() + projectileDelay;

	const spawnAnimation = new TimelineComponent(target);
	spawnAnimation.registerComponent();
	spawnAnimation.setTimelineLength(0.7);
	spawnAnimation.addFloatCurve(trackName, target.spawnAnimationCurve);

	spawnAnimation.setOnFloatUpdate(trackName, value => {
		onAnimate(target, value);
	});

	spawnAnimation.setOnFinished(() => {
		target.movementEnabled = true;
	});

	spawnAnimation.playFromStart();

	target.addOwnedComponent(spawnAnimation);
}

function onAnimate(target: BP_PixelShipEnemy, value: number) {
	const { spawnLocation, gameCabinetYaw } = target;

	const sourceLocation = spawnLocation;
	const targetLocation = sourceLocation.subtract(new Vector(0, 0, 60));
	const location = KMath.vlerp(sourceLocation, targetLocation, value);
	target.setActorLocation(location);

	const sourceRotation = new Rotator(0, 0, gameCabinetYaw + 180);
	const targetRotation = new Rotator(0, 0, gameCabinetYaw);
	const rotation = KMath.rlerp(sourceRotation, targetRotation, value, false);
	target.setActorRotation(rotation);
}

export function tick(target: BP_PixelShipEnemy, deltaSeconds: number) {
	tryMove(target, deltaSeconds);
	tryFireProjectile(target);
}

function tryMove(target: BP_PixelShipEnemy, deltaSeconds: number) {
	if (!target.movementEnabled) { return; }

	const speedSign = target.movementDirection ? 1 : -1;
	const speed = target.movementSpeed * speedSign;
	const cabinetRotation = new Rotator(0, 0, target.gameCabinetYaw);
	const velocity = cabinetRotation.rotateVector(new Vector(speed, 0, -45));
	const offset = velocity.multiplyFloat(deltaSeconds);
	target.addActorLocalOffset(offset, true);
}

function tryFireProjectile(target: BP_PixelShipEnemy) {
	if (!target.movementEnabled) { return; }

	const currentTime = KSystem.getGameTimeInSeconds();
	if (currentTime < target.nextProjectileTime) { return; }

	const pawn = GameplayStatics.getPlayerPawn(0);
	const pawnLocation = pawn.getActorLocation();
	const targetLocation = target.getActorLocation();
	const targetToPawn = pawnLocation.subtract(targetLocation);
	if (targetToPawn.x > 120) { return; }

	target.spawnProjectile(
		target.getTransform(),
		new Vector(0, 0, -200),
		target,
		KMath.makeColor(1, 0.3, 0.05)
	);

	const projectileDelay = KMath.randomFloatInRange(0.4, 0.8);
	target.nextProjectileTime = currentTime + projectileDelay;
}

export function onHit(target: BP_PixelShipEnemy, other: Actor) {
	if (other instanceof Pawn) {
		other.applyDamage(90, null, null, null);
		destroy(target);
	} else {
		target.movementDirection = !target.movementDirection;
	}
}

export function onDamage(target: BP_PixelShipEnemy, damage: number) {
	target.health -= damage;

	if (target.health <= 0) {
		destroy(target);
	}
}

function destroy(target: BP_PixelShipEnemy) {
	GameplayStatics.spawnEmitterAtLocation(
		target.destroyedEffect,
		target.getActorLocation(),
		new Rotator(0, 0, 0),
		new Vector(1, 1, 1)
	);

	target.destroyActor();
}
